### ISSUE TITLE
min_delta does not function correctly

### ISSUE THREAD

**Issue by jwarren1984** [fchollet/keras#4360]
It's not clear what `min_delta` is meant to do; should it be cumulative (i.e., the cumulative change in the monitor value over the past `patience` epochs not exceeding `min_delta`) or not (an individual change being less than `min_delta`)? In either case, it doesn't seem to work properly. 

```
if self.monitor_op(current - self.min_delta, self.best):
            self.best = current
            self.wait = 0

```
As you can see, if `current ` is simply staying around `self.min_delta` above `self.best` , then the conditional is satisfied and `self.wait` is reset to zero and the thing never stops. Also, it appears that `self.best` can be reset to `current` even if `current` is worse than `self.best`, as long as it's within `self.min_delta` of `self.best`. This is consistent with my experience, and I can't imagine that this is the intended behavior. If I am mistaken, I apologize for raising this issue.

I think it would make more sense to use some sort of relative tolerance over a window of epochs, such that if the percentage change in the metric over that window is less than some tolerance value the training stops. Thoughts?

**Comment by jwarren1984** [comment]
Here's how I implemented a relative tolerance. `Self.wait` is now a log of the past `patience` monitor values. The program checks to see if the average percent improvement over the past `patience` epochs exceeds the `min_delta`; if not, it stops training.

```
class EarlyStopping(Callback):
    '''Stop training when a monitored quantity has stopped improving.
    # Arguments
        monitor: quantity to be monitored.
        min_delta: minimum change in the monitored quantity
            to qualify as an improvement, i.e. an absolute
            change of less than min_delta, will count as no
            improvement.
        patience: number of epochs with no improvement
            after which training will be stopped.
        verbose: verbosity mode.
        mode: one of {auto, min, max}. In `min` mode,
            training will stop when the quantity
            monitored has stopped decreasing; in `max`
            mode it will stop when the quantity
            monitored has stopped increasing; in `auto`
            mode, the direction is automatically inferred
            from the name of the monitored quantity.
    '''

    def __init__(self, monitor='val_loss', min_delta=0, patience=0, verbose=0, mode='auto'):
        super(EarlyStopping, self).__init__()

        self.monitor = monitor
        self.patience = patience
        self.verbose = verbose
        self.min_delta = min_delta
        # self.wait = 0
        self.wait = []
        self.stopped_epoch = 0

        if mode not in ['auto', 'min', 'max']:
            warnings.warn('EarlyStopping mode %s is unknown, '
                          'fallback to auto mode.' % (self.mode),
                          RuntimeWarning)
            mode = 'auto'

        if mode == 'min':
            self.monitor_op = np.less
        elif mode == 'max':
            self.monitor_op = np.greater
        else:
            if 'acc' in self.monitor:
                self.monitor_op = np.greater
            else:
                self.monitor_op = np.less

        if self.monitor_op == np.greater:
            self.min_delta *= 1
        else:
            self.min_delta *= -1

    def on_train_begin(self, logs={}):
        # self.wait = 0  # Allow instances to be re-used
        self.wait = []
        # self.best = np.Inf if self.monitor_op == np.less else -np.Inf

    def on_epoch_end(self, epoch, logs={}):
        current = logs.get(self.monitor)
        if current is None:
            warnings.warn('Early stopping requires %s available!' %
                          (self.monitor), RuntimeWarning)

        if len(self.wait) < self.patience:
            self.wait = np.append(self.wait, current)
        else:
            self.wait = np.roll(self.wait, -1)
            self.wait[-1] = current
            rel_improvement = np.mean(np.diff(self.wait)) / np.mean(self.wait)
            if not self.monitor_op(rel_improvement, self.min_delta):
                self.stopped_epoch = epoch
                self.model.stop_training = True
        # if self.monitor_op(current - self.min_delta, self.best):
        #     self.best = current
        #     self.wait = 0
        # else:
        #     if self.wait >= self.patience:
        #         self.stopped_epoch = epoch
        #         self.model.stop_training = True
        #     self.wait += 1

    def on_train_end(self, logs={}):
        if self.stopped_epoch > 0 and self.verbose > 0:
            print('Epoch %05d: early stopping' % (self.stopped_epoch))
```

**Comment by jwarren1984** [comment]
It's better to have `rel_improvement` reflect improvement per epoch (otherwise the value of `min_delta` can have vastly different effects for different values of `patience`). Therefore, divide by `patience `when computing `rel_improvement`, as shown below.

```
class EarlyStopping(Callback):
    '''Stop training when a monitored quantity has stopped improving.
    # Arguments
        monitor: quantity to be monitored.
        min_delta: minimum change in the monitored quantity
            to qualify as an improvement, i.e. an absolute
            change of less than min_delta, will count as no
            improvement.
        patience: number of epochs with no improvement
            after which training will be stopped.
        verbose: verbosity mode.
        mode: one of {auto, min, max}. In `min` mode,
            training will stop when the quantity
            monitored has stopped decreasing; in `max`
            mode it will stop when the quantity
            monitored has stopped increasing; in `auto`
            mode, the direction is automatically inferred
            from the name of the monitored quantity.
    '''

    def __init__(self, monitor='val_loss', min_delta=0, patience=0, verbose=0, mode='auto'):
        super(EarlyStopping, self).__init__()

        self.monitor = monitor
        self.patience = patience
        self.verbose = verbose
        self.min_delta = min_delta
        # self.wait = 0
        self.wait = []
        self.stopped_epoch = 0

        if mode not in ['auto', 'min', 'max']:
            warnings.warn('EarlyStopping mode %s is unknown, '
                          'fallback to auto mode.' % (self.mode),
                          RuntimeWarning)
            mode = 'auto'

        if mode == 'min':
            self.monitor_op = np.less
        elif mode == 'max':
            self.monitor_op = np.greater
        else:
            if 'acc' in self.monitor:
                self.monitor_op = np.greater
            else:
                self.monitor_op = np.less

        if self.monitor_op == np.greater:
            self.min_delta *= 1
        else:
            self.min_delta *= -1

    def on_train_begin(self, logs={}):
        # self.wait = 0  # Allow instances to be re-used
        self.wait = []
        # self.best = np.Inf if self.monitor_op == np.less else -np.Inf

    def on_epoch_end(self, epoch, logs={}):
        current = logs.get(self.monitor)
        if current is None:
            warnings.warn('Early stopping requires %s available!' %
                          (self.monitor), RuntimeWarning)

        if len(self.wait) < self.patience:
            self.wait = np.append(self.wait, current)
        else:
            self.wait = np.roll(self.wait, -1)
            self.wait[-1] = current
            rel_improvement = np.mean(np.diff(self.wait)) / np.mean(self.wait) / self.patience
            if not self.monitor_op(rel_improvement, self.min_delta):
                self.stopped_epoch = epoch
                self.model.stop_training = True
        # if self.monitor_op(current - self.min_delta, self.best):
        #     self.best = current
        #     self.wait = 0
        # else:
        #     if self.wait >= self.patience:
        #         self.stopped_epoch = epoch
        #         self.model.stop_training = True
        #     self.wait += 1

    def on_train_end(self, logs={}):
        if self.stopped_epoch > 0 and self.verbose > 0:
            print('Epoch %05d: early stopping' % (self.stopped_epoch))
```
